### ISSUE TITLE
FR: remove need for dedicated class when running same task twice with different configuration

### ISSUE THREAD

**Issue by keradus** [phpro/grumphp#501]
docs refs: https://github.com/phpro/grumphp/pull/483/files#diff-ccf4f1a1622265fd2805632c80129a8bR193
original request: https://github.com/phpro/grumphp/issues/309#issuecomment-377535886

I'm opening new ticket, as one in which it was requested is closed - original ticket was about documenting status quo, this one is to bring better User eXerience.

This gonna be super useful for running different shell tasks, which obviously has same implementation, just different configuration, and making new dummy-class for every single shell-task is way not friendly

**Comment by keradus** [comment]
> Since we currently match the name based on the task name

@veewee , can you point place of code when this is happening, and maybe propose direction how it could be changed ?

**Comment by veewee** [comment]
Hi @keradus,

Sorry for the late response.
I can point you to following locations:

https://github.com/phpro/grumphp/blob/master/doc/tasks.md#creating-a-custom-task
This information is on how the tasks are configured / registered in the service container. 
You can see that the task service contains a config key. This one is linked through a compiler pass.

https://github.com/phpro/grumphp/blob/master/src/Configuration/Compiler/TaskCompilerPass.php
This is the compiler pass that is responsible for registering the task in the Taskrunner, normalizing the configuration and storing it into new variables inside the service container. Those variables can be accessed by the `GrumPHP` configuration object at: https://github.com/phpro/grumphp/blob/master/src/Configuration/GrumPHP.php

https://github.com/phpro/grumphp/blob/master/src/Task/TaskInterface.php#L14
Finally, there is the `getName()` method inside the task which is being used in the `TaskRunner` to filter, report, fetch additional configutation, ...


Currently, I do not have a direction proposition on the issue. Since the name is used in 3 locations, I think a first phase would be to make it easier to configure and register tasks internally. 
Another thing I am thinking of, is to make it possible to create an aliases system in the configuration. However: since the `Task::getName()` is being used inside the reporting, this might require a BC break inside the `TaskInterface` - which I would like to avoid at the moment.

I hope this gives you a better understanding on the issue. Feel free to ask additional information or to propose a plan of action.

**Comment by keradus** [comment]
indeed, looks like a bigger thing indeed 

thanks for detailed description, I'm not feeling able to do it myself right now, but let us have this FR open please ;) it's not first time it's requested, and now we have more info than single-line comment in src why one has to create custom task.

**Comment by RaphTbm** [comment]
Hi there,

Not sure if it's really possible without a BC break.

Here my proposal (it's working on my environment):

In grumphp.yml
````yml
parameters:
  tasks:
    shell:
      scripts:
        - scripts/ci/basic.sh
      triggered_by: [php]
      extra:
        label: 'Just Shell'
    shell.sass:
      scripts:
        - scripts/ci/sass-lint.sh
      triggered_by: [scss]
      extra:
        label: 'Shell Sass'

services:
  task.shell.sass:
    parent: task.shell
    tags:
      - {name: grumphp.task, config: shell.sass}
````

In TaskCompilerPass
````php
$definition->addMethodCall('addTask', [new Reference($id), $configKey]);
````

In TaskRunner
````php
/**
 * @param TaskInterface $task
 * @param string $configKey
 */
public function addTask(TaskInterface $task, $configKey)
{
    if ($this->tasks->contains($task)) {
        return;
    }
    $task->setName($configKey);
    $this->tasks->add($task);
}
````

But it means, we need to remove getName from all the current taks and implements something like that in the necessary abstractClass, example in AbstractExternalTask:
````php
/**
 * @return string
 */
public function getName()
{
    return $this->name;
}

public function setName($name) {
    $this->name = $name;
}
````

And all the task extending directly the interface should extend of one the abstract class existing instead (as FileSize.php, ...).

The extra key, is an extra suggestion related to this issue, we should allow the possibility to use some custom label. Dirty example to illustrate my idea:

In AbstracExternalTask
````php
/**
 * @return string
 */
public function getExtraOption($key)
{
    if (!$this->extraOptions && $this->extraOptions !== FALSE) {
        $this->extraOptions = FALSE;
        $configured = $this->grumPHP->getTaskConfiguration($this->getName());
        if (isset($configured['extra'])) {
            $this->extraOptions = $configured['extra'];
        }
    }

    return $this->extraOptions && isset($this->extraOptions[$key]) ? $this->extraOptions[$key] : NULL;
}
````

In ProgressSubscriber
````php
/**
 * @param TaskEvent $event
 */
public function advanceProgress(TaskEvent $event)
{
    $task = $event->getTask();
    if (!$taskName = $task->getExtraOption('label')) {
        $taskReflection = new ReflectionClass($event->getTask());
        $taskName = $taskReflection->getShortName();
    }

    $this->progressBar->setFormat($this->progressFormat);
    $this->progressBar->setMessage($taskName);
    $this->progressBar->advance();
}
````

**Comment by veewee** [comment]
Fixed in #711
